### ISSUE TITLE
Net core version much slower?

### ISSUE THREAD

**Issue by PatriceDargenton** [jdermody/brightwire#47]
Hello, I have tried both versions of your engine, they work fine, but the net core version is very much slower, here is my test:

    /*
    <Project Sdk="Microsoft.NET.Sdk">
      <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net5.0</TargetFramework>
      </PropertyGroup>
      <ItemGroup>
        <PackageReference Include="BrightData.Numerics" Version="3.0.2" />
        <PackageReference Include="BrightWire" Version="3.0.2" />
      </ItemGroup>
    </Project>
    */

    using BrightData; // BrightDataContext
    using BrightWire; // CreateGraphFactory
    using BrightWire.Models; // ExecutionGraphModel
    using System; // DateTime
    using System.Linq; // ToList
    using BrightData.Numerics; // UseNumericsLinearAlgebra

    namespace ConsoleAppNetCore
    {
        public static class Program
        {
            public static void Main(string[] args)
            {
                XORTestNetCore();
            }

            public static void XORTestNetCore()
            {
                // determinist
                //const int RandomSeed = 1;
                //using var context = new BrightDataContext(randomSeed: RandomSeed);
                using var context = new BrightDataContext(randomSeed: null); // stochastic

                context.UserNotifications = null; // Disable user notifications
            
                context.ResetRandom(seed: null); // stochastic
                //context.ResetRandom(seed: RandomSeed); // determinist

                context.UseNumericsLinearAlgebra();

                var builder = context.BuildTable();
                builder.AddColumn(BrightData.BrightDataType.Vector, "Input");
                builder.AddColumn(BrightData.BrightDataType.Vector, "Outpout").SetTarget(true);

                var doubleXOR = true;

                if (!doubleXOR) { // 1 XOR
                    builder.AddRow(context.CreateVector(1.0f, 0.0f), context.CreateVector(1.0f));
                    builder.AddRow(context.CreateVector(0.0f, 0.0f), context.CreateVector(0.0f));
                    builder.AddRow(context.CreateVector(0.0f, 1.0f), context.CreateVector(1.0f));
                    builder.AddRow(context.CreateVector(1.0f, 1.0f), context.CreateVector(0.0f));
                } else { // 2 XOR
                    builder.AddRow(context.CreateVector(1.0f, 0.0f, 1.0f, 0.0f), context.CreateVector(1.0f, 1.0f)); // 1 0 1 0  1 1
                    builder.AddRow(context.CreateVector(1.0f, 0.0f, 0.0f, 0.0f), context.CreateVector(1.0f, 0.0f)); // 1 0 0 0  1 0
                    builder.AddRow(context.CreateVector(1.0f, 0.0f, 0.0f, 1.0f), context.CreateVector(1.0f, 1.0f)); // 1 0 0 1  1 1
                    builder.AddRow(context.CreateVector(1.0f, 0.0f, 1.0f, 1.0f), context.CreateVector(1.0f, 0.0f)); // 1 0 1 1  1 0
                    builder.AddRow(context.CreateVector(0.0f, 0.0f, 1.0f, 0.0f), context.CreateVector(0.0f, 1.0f)); // 0 0 1 0  0 1
                    builder.AddRow(context.CreateVector(0.0f, 0.0f, 0.0f, 0.0f), context.CreateVector(0.0f, 0.0f)); // 0 0 0 0  0 0
                    builder.AddRow(context.CreateVector(0.0f, 0.0f, 0.0f, 1.0f), context.CreateVector(0.0f, 1.0f)); // 0 0 0 1  0 1
                    builder.AddRow(context.CreateVector(0.0f, 0.0f, 1.0f, 1.0f), context.CreateVector(0.0f, 0.0f)); // 0 0 1 1  0 0
                    builder.AddRow(context.CreateVector(0.0f, 1.0f, 1.0f, 0.0f), context.CreateVector(1.0f, 1.0f)); // 0 1 1 0  1 1
                    builder.AddRow(context.CreateVector(0.0f, 1.0f, 0.0f, 0.0f), context.CreateVector(1.0f, 0.0f)); // 0 1 0 0  1 0
                    builder.AddRow(context.CreateVector(0.0f, 1.0f, 0.0f, 1.0f), context.CreateVector(1.0f, 1.0f)); // 0 1 0 1  1 1
                    builder.AddRow(context.CreateVector(0.0f, 1.0f, 1.0f, 1.0f), context.CreateVector(1.0f, 0.0f)); // 0 1 1 1  1 0
                    builder.AddRow(context.CreateVector(1.0f, 1.0f, 1.0f, 0.0f), context.CreateVector(0.0f, 1.0f)); // 1 1 1 0  0 1
                    builder.AddRow(context.CreateVector(1.0f, 1.0f, 0.0f, 0.0f), context.CreateVector(0.0f, 0.0f)); // 1 1 0 0  0 0
                    builder.AddRow(context.CreateVector(1.0f, 1.0f, 0.0f, 1.0f), context.CreateVector(0.0f, 1.0f)); // 1 1 0 1  0 1
                    builder.AddRow(context.CreateVector(1.0f, 1.0f, 1.0f, 1.0f), context.CreateVector(0.0f, 0.0f)); // 1 1 1 1  0 0
                }

                var training = builder.BuildRowOriented();
                var test = training;

                var graph = context.CreateGraphFactory();

                graph.CurrentPropertySet.
                    Use(graph.GaussianWeightInitialisation(
                        zeroBias: true, stdDev: 0.1f,
                        varianceCalibration: GaussianVarianceCalibration.SquareRoot2N));

                graph.CurrentPropertySet.Use(graph.GradientDescent.RmsProp);

                var errorMetric = graph.ErrorMetric.BinaryClassification;

                var traininData = graph.CreateDataSource(test);
                var testData = traininData.CloneWith(test);
                var grapheTrainingEngine = graph.CreateTrainingEngine(traininData, errorMetric,
                    learningRate: 0.1f, batchSize: 100);
                uint hiddenLayerSize = 12;
                var outputSize = grapheTrainingEngine.DataSource.GetOutputSizeOrThrow();
                graph.Connect(grapheTrainingEngine)
                    .AddFeedForward(hiddenLayerSize)
                    .Add(graph.TanhActivation())
                    .AddFeedForward(outputSize)
                    .Add(graph.TanhActivation())
                    .AddBackpropagation();

                var timeStart = DateTime.Now;
                uint numIterations = 2000;
                ExecutionGraphModel bestGraph = null;
                grapheTrainingEngine.Train(numIterations, testData, model => bestGraph = model.Graph);

                if (bestGraph != null)
                {
                    var executionEngine = graph.CreateExecutionEngine(bestGraph);
                    var output = executionEngine.Execute(testData).ToList();
                    //uint outputSize = (uint)(executionEngine.DataSource.OutputSize);
                    var testAccuracy = output.Average(
                        o => o.CalculateError(graph.ErrorMetric.OneHotEncoding));
                    Console.WriteLine($"Neural network accuracy: {testAccuracy:P}");
                    foreach (var item in output)
                    {
                        foreach (var index in item.MiniBatchSequence.MiniBatch.Rows)
                        {
                            var row = test.Row(index);
                            var result = item.Output[index];
                            if (outputSize == 1)
                                Console.WriteLine($"{row} = {result} : " + 
                                    result[0].ToString("0.00"));
                            if (outputSize == 2)
                                Console.WriteLine($"{row} = {result} : " +
                                    result[0].ToString("0.00") + ", " +
                                    result[1].ToString("0.00"));
                        }
                    }
                }
                var timeEnd = DateTime.Now;
                var ts = timeEnd - timeStart;
                Console.WriteLine("Duration sec.: " + ts.TotalSeconds);

                System.Diagnostics.Debug.WriteLine("!");
            }
        }
    }

	/*
	<packages>
	  <package id="BrightWire.Net4" version="2.1.1" targetFramework="net472" />
	  <package id="MathNet.Numerics" version="4.7.0" targetFramework="net472" />
	  <package id="protobuf-net" version="2.4.0" targetFramework="net472" />
	  <package id="System.ValueTuple" version="4.5.0" targetFramework="net472" />
	</packages>
	*/

	using BrightWire; // BrightWireProvider
	using BrightWire.ExecutionGraph; // BrightWireProvider
	using System; // DateTime
	using System.Linq; // Average

	namespace ConsoleAppNet4
	{
		class Program
		{
			static void Main(string[] args)
			{
				XORTestNet4();
				Console.ReadKey();
			}

			public static void XORTestNet4()
			{
				using (var lap = BrightWireProvider.CreateLinearAlgebra(stochastic:true))
				{
					var builder = BrightWireProvider.CreateDataTableBuilder();

					var doubleXOR = true;

					if (!doubleXOR) { 
						builder.AddVectorColumn(2, "Input");
						builder.AddVectorColumn(1, "Output", isTarget: true);
						var fv1 = new BrightWire.Models.FloatVector();
						var fv2 = new BrightWire.Models.FloatVector();
						var fv3 = new BrightWire.Models.FloatVector();
						var fv4 = new BrightWire.Models.FloatVector();
						var fvi1 = new BrightWire.Models.FloatVector();
						var fvi2 = new BrightWire.Models.FloatVector();
						var fvi3 = new BrightWire.Models.FloatVector();
						var fvi4 = new BrightWire.Models.FloatVector();
						fvi1.Data = new[] { 1.0f, 0.0f }; fv1.Data = new[] { 1.0f }; builder.Add(fvi1, fv1); // 1 0  1
						fvi2.Data = new[] { 0.0f, 0.0f }; fv2.Data = new[] { 0.0f }; builder.Add(fvi2, fv2); // 0 0  0
						fvi3.Data = new[] { 0.0f, 1.0f }; fv3.Data = new[] { 1.0f }; builder.Add(fvi3, fv3); // 0 1  1
						fvi4.Data = new[] { 1.0f, 1.0f }; fv4.Data = new[] { 0.0f }; builder.Add(fvi4, fv4); // 1 1  0
					}
					else {

						builder.AddVectorColumn(4, "Input");
						builder.AddVectorColumn(2, "Output", isTarget: true);

						var fv1 = new BrightWire.Models.FloatVector();
						var fv2 = new BrightWire.Models.FloatVector();
						var fv3 = new BrightWire.Models.FloatVector();
						var fv4 = new BrightWire.Models.FloatVector();
						var fv5 = new BrightWire.Models.FloatVector();
						var fv6 = new BrightWire.Models.FloatVector();
						var fv7 = new BrightWire.Models.FloatVector();
						var fv8 = new BrightWire.Models.FloatVector();
						var fv9 = new BrightWire.Models.FloatVector();
						var fv10 = new BrightWire.Models.FloatVector();
						var fv11 = new BrightWire.Models.FloatVector();
						var fv12 = new BrightWire.Models.FloatVector();
						var fv13 = new BrightWire.Models.FloatVector();
						var fv14 = new BrightWire.Models.FloatVector();
						var fv15 = new BrightWire.Models.FloatVector();
						var fv16 = new BrightWire.Models.FloatVector();

						var fvi1 = new BrightWire.Models.FloatVector();
						var fvi2 = new BrightWire.Models.FloatVector();
						var fvi3 = new BrightWire.Models.FloatVector();
						var fvi4 = new BrightWire.Models.FloatVector();
						var fvi5 = new BrightWire.Models.FloatVector();
						var fvi6 = new BrightWire.Models.FloatVector();
						var fvi7 = new BrightWire.Models.FloatVector();
						var fvi8 = new BrightWire.Models.FloatVector();
						var fvi9 = new BrightWire.Models.FloatVector();
						var fvi10 = new BrightWire.Models.FloatVector();
						var fvi11 = new BrightWire.Models.FloatVector();
						var fvi12 = new BrightWire.Models.FloatVector();
						var fvi13 = new BrightWire.Models.FloatVector();
						var fvi14 = new BrightWire.Models.FloatVector();
						var fvi15 = new BrightWire.Models.FloatVector();
						var fvi16 = new BrightWire.Models.FloatVector();

						fvi1.Data = new[] { 1.0f, 0.0f, 1.0f, 0.0f }; fv1.Data = new[] { 1.0f, 1.0f }; builder.Add(fvi1, fv1);     // 1 0 1 0  1 1
						fvi2.Data = new[] { 1.0f, 0.0f, 0.0f, 0.0f }; fv2.Data = new[] { 1.0f, 0.0f }; builder.Add(fvi2, fv2);     // 1 0 0 0  1 0
						fvi3.Data = new[] { 1.0f, 0.0f, 0.0f, 1.0f }; fv3.Data = new[] { 1.0f, 1.0f }; builder.Add(fvi3, fv3);     // 1 0 0 1  1 1
						fvi4.Data = new[] { 1.0f, 0.0f, 1.0f, 1.0f }; fv4.Data = new[] { 1.0f, 0.0f }; builder.Add(fvi4, fv4);     // 1 0 1 1  1 0
						fvi5.Data = new[] { 0.0f, 0.0f, 1.0f, 0.0f }; fv5.Data = new[] { 0.0f, 1.0f }; builder.Add(fvi5, fv5);     // 0 0 1 0  0 1
						fvi6.Data = new[] { 0.0f, 0.0f, 0.0f, 0.0f }; fv6.Data = new[] { 0.0f, 0.0f }; builder.Add(fvi6, fv6);     // 0 0 0 0  0 0
						fvi7.Data = new[] { 0.0f, 0.0f, 0.0f, 1.0f }; fv7.Data = new[] { 0.0f, 1.0f }; builder.Add(fvi7, fv7);     // 0 0 0 1  0 1
						fvi8.Data = new[] { 0.0f, 0.0f, 1.0f, 1.0f }; fv8.Data = new[] { 0.0f, 0.0f }; builder.Add(fvi8, fv8);     // 0 0 1 1  0 0
						fvi9.Data = new[] { 0.0f, 1.0f, 1.0f, 0.0f }; fv9.Data = new[] { 1.0f, 1.0f }; builder.Add(fvi9, fv9);     // 0 1 1 0  1 1
						fvi10.Data = new[] { 0.0f, 1.0f, 0.0f, 0.0f }; fv10.Data = new[] { 1.0f, 0.0f }; builder.Add(fvi10, fv10); // 0 1 0 0  1 0
						fvi11.Data = new[] { 0.0f, 1.0f, 0.0f, 1.0f }; fv11.Data = new[] { 1.0f, 1.0f }; builder.Add(fvi11, fv11); // 0 1 0 1  1 1
						fvi12.Data = new[] { 0.0f, 1.0f, 1.0f, 1.0f }; fv12.Data = new[] { 1.0f, 0.0f }; builder.Add(fvi12, fv12); // 0 1 1 1  1 0
						fvi13.Data = new[] { 1.0f, 1.0f, 1.0f, 0.0f }; fv13.Data = new[] { 0.0f, 1.0f }; builder.Add(fvi13, fv13); // 1 1 1 0  0 1
						fvi14.Data = new[] { 1.0f, 1.0f, 0.0f, 0.0f }; fv14.Data = new[] { 0.0f, 0.0f }; builder.Add(fvi14, fv14); // 1 1 0 0  0 0
						fvi15.Data = new[] { 1.0f, 1.0f, 0.0f, 1.0f }; fv15.Data = new[] { 0.0f, 1.0f }; builder.Add(fvi15, fv15); // 1 1 0 1  0 1
						fvi16.Data = new[] { 1.0f, 1.0f, 1.0f, 1.0f }; fv16.Data = new[] { 0.0f, 0.0f }; builder.Add(fvi16, fv16); // 1 1 1 1  0 0
					}

					var dataTable = builder.Build();

					var graph = new GraphFactory(lap);
					var errorMetric = graph.ErrorMetric.BinaryClassification;
					graph.CurrentPropertySet
						.Use(graph.GradientDescent.RmsProp)
						.Use(graph.WeightInitialisation.Gaussian);

					var testData = graph.CreateDataSource(dataTable);
					var engine = graph.CreateTrainingEngine(testData, 
						learningRate: 0.1f, batchSize: 100);
					var outputSize = engine.DataSource.OutputSize;

					const int HIDDEN_LAYER_SIZE = 12;
					graph.Connect(engine)
						.AddFeedForward(HIDDEN_LAYER_SIZE)
						.Add(graph.TanhActivation())
						.AddFeedForward(engine.DataSource.OutputSize)
						.Add(graph.TanhActivation())
						.AddBackpropagation(errorMetric);

					var timeStart = DateTime.Now;
					var executionContext = graph.CreateExecutionContext();
					for (var i = 0; i < 2000; i++)
					{
						engine.Train(executionContext);
						if (i % 100 == 0)
							engine.Test(testData, errorMetric);
					}
					engine.Test(testData, errorMetric);

					var networkGraph = engine.Graph;
					var executionEngine = graph.CreateEngine(networkGraph);
					var output = executionEngine.Execute(testData);
					Console.WriteLine(output.Average(o => o.CalculateError(errorMetric)));

					foreach (var item in output)
					{
						foreach (var index in item.MiniBatchSequence.MiniBatch.Rows)
						{
							var row = dataTable.GetRow(index);
							var result = item.Output[index];
							if (outputSize == 1)
								Console.WriteLine(
									row.Data[0] + " XOR " +
									row.Data[1] + " : " + 
									result.Data[0].ToString("0.00"));
							if (outputSize == 2)
							{
								var fv = (BrightWire.Models.FloatVector)row.Data[0];
								Console.WriteLine(
									fv.Data[0] + " XOR " +
									fv.Data[1] + ", " +
									fv.Data[2] + " XOR " +
									fv.Data[3] + " : " +
									result.Data[0].ToString("0.00") + " : " +
									result.Data[1].ToString("0.00"));
							}
						}
					}
				
					var timeEnd = DateTime.Now;
					var ts = timeEnd - timeStart;
					Console.WriteLine("Duration sec.: " + ts.TotalSeconds);

					System.Diagnostics.Debug.WriteLine("!");
				}
			}
		}
	}

**Comment by jdermody** [comment]
I think the two tests are not quite the same - in the .net core version the following line:

`grapheTrainingEngine.Train(numIterations, testData, model => bestGraph = model.Graph);`

is not the same as:

```
for (var i = 0; i < 2000; i++)
{
	engine.Train(executionContext);
	if (i % 100 == 0)
		engine.Test(testData, errorMetric);
}
```

To reproduce the same test case the .net core version should be changed to:

`grapheTrainingEngine.Train(numIterations, testData, model => bestGraph = model.Graph**, 100**);`

For me this means that the .net core version becomes ~300ms vs ~200ms for .net 4 version, admittedly still slower but not by so much.

**Comment by PatriceDargenton** [comment]
Ok, thank you very much, now all of my functional tests are running at almost the same speed!
